mail-filter/spamassassin: GLEP-81 compliance

Philippe23 opened the pull request:
mail-filter/spamassassin: GLEP-81 compliance

Closes: https://github.com/gentoo/gentoo/pull/14055
Package-Manager: Portage-2.3.76, Repoman-2.3.16
Signed-off-by: Philippe Chaintreuil <gentoo_bugs_peep@parallaxshift.com>

Philippe23 commented:
@orlitzky: This is the v3.4.3 version of that earlier PR[1].  I just remembered I need to add that spamd home migration notification, so you can just ignore this until I update this PR and ping you again.

[1] https://github.com/gentoo/gentoo/pull/13766

Philippe23 commented:
@orlitzky:

Added reminder.  Let me know if you have any suggestions/tweaks.

- I'm only outputting it when /home/spamd exists as a proxy for the old setup.
- On my first dry run, the acct-user eclass warned that it was unable to change the spamd user's home folder, so I added the suggestion to run `usermod -d`
- Not sure if there's a better way than what's suggested.
- I thought I saw some mention of a master file for keeping track of uid/gid's, but I'm not sure where that is, so it's not in this changelist.

No rush on this.  Hope the holidays are treating you well.

Philippe23 commented:
Found uid-gid.txt and submitted PR on that repo: https://github.com/gentoo/api-gentoo-org/pull/238

orlitzky commented:
I finally posted a message to -dev asking if we should be allowed to use `/home/spamd`. That would make this upgrade a lot less annoying for everyone.

orlitzky commented:
The discussion on -dev was... not productive. Here's what I've gathered:

 * `/home/spamd` is the most appropriate place for this home directory, since it stores user-specific data created by a human.
 * Leaving it set to `/home/spamd` will make the migration a lot easier, and won't immediately break the mail servers of everyone who upgrades.
 * QA isn't going to fix the obsolete warning.
 * We can't work around the warning in the ebuild because the eclass is rigged to `die()` if you delete the stuff that triggers the warning in `src_install`.
    * If you delete the stuff that triggers the warning _later_, then the warning still gets shown for a file that doesn't exist!

So ultimately, the correct and easy thing to do here is just set it to `/home/spamd` and ignore the resulting QA warning about the dummy keepdir file that gets installed. It's stupid, but it's self-inflicted. The warning is wrong and someone will fix it one day when it annoys them and not us.

Philippe23 commented:
I've just pushed changes to revert to /home/spamd (and removed the block of text).  Permissions of /home/spamd are still tightened.  I've switched my machine back from /var/lib/spamd to /home/spamd.  Both worked for me, but I'm not using amavis etc.

Footnote: I read through todays -dev digest, and got lost a little at times when "$HOME" was being used as to whether it was meaning ~spamd or ~someuser.  

When we submit this, you may wish to consider submitting the bug about the QA warning yourself to document why it's there.

orlitzky commented:
Thanks, I've spent hours arguing about it so I'm not likely to forget.

I'm going to write a patch for the post-install check to exclude `/home/${PN}/.keep*` files for acct-user packages, and I'll merge this after I've posted that to the ML.

orlitzky commented:
One more request, sorry: can you please add the group and user packages in separate commits (group then user)?

Philippe23 commented:
No worries.  But a few questions:

1.)  Fine if they're all part of this same PR?  Or are you after three separate PRs?
2.)  Is this a personal preference thing, or is there a logistical reason?  (Curious for my own education if it's the latter.)

orlitzky commented:
They can all be a part of the same PR, it's easier that way.

Gentoo does try to limit commits to a single package. Tools like `repoman commit` are designed to handle commits only to one package; for example it prepends "category/package: " to your commit message, and runs "repoman scan" on the package that lives in the current directory. Both of those only work one-package-at-a-time.

Even though those are just "implementation details," the end result agrees with my own personal preference. If a change is self-contained, I prefer to make it in a separate commit because it makes the git log easier to read (you're reading a commit message about one thing instead of five), and it makes that one change easy to revert on its own if things break.

In Gentoo, the unit of "self-contained" changes is pretty much by design the package, which is why the two agree. For example you could add the group package first, run repoman etc., and the tree would be perfectly consistent. Then when you add the user package, it only depends on the group package, so its dependencies are satisfied and the tree is consistent. If you pushed those changes individually and walked away for a week, everything would be fine.

On the other hand, if you tried to commit only a metadata.xml file without an ebuild, repoman would throw a fit and portage might break, so that's not a self-contained change. Ideally, you can check out any commit from the tree and be able to work with it... ideally.

Philippe23 commented:
Split up.

NOTE: I added a comment about `/home/spamd` vs QA checks in `acct-user/spamd`, otherwise it's the same contents.

orlitzky commented:
On second thought, I no longer have the mental energy to start every day pissed off from arguing with strangers on the internet. The QA team is still giving me a hard time over this, so I've committed this as-is to our company overlay and have removed myself as a maintainer from spamassassin. I'm sorry to leave you hanging -- you've been a great proxy maintainer -- but I don't want to do volunteer work that makes me miserable any more.

Philippe23 commented:
@orlitzky: Thanks for the help and guidance over the years.  Sad to see you step back, but I think it sounds like the right move for you.  I'll keep trudging along.  I hope you'll remain a user who throws in your opinion as time goes on, since I value your input.

Philippe23 commented:
Notes to self: Amavis has already started figuring out the /home/spamd vs /var/lib/spamd questions.  Here's a [comment from mjo describing some of the issues](https://github.com/gentoo/gentoo/pull/13265#issuecomment-543130029).  And here's them [confirming that /var/lib/spamd/home won't work](https://github.com/gentoo/gentoo/pull/13265#issuecomment-575923240).

orlitzky commented:
Sibling home directory is explicitly forbidden by [the FHS](http://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html):

> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. Users must never need to modify files in /var/lib to configure a package's operation, and the specific file hierarchy used to store the data must not be exposed to regular users.

Specifically, both of these are problems:

* State information is data that programs modify while they run
* Users must never need to modify files in /var/lib to configure a package's operation

But I'm sorry for butting in after I said I wouldn't. I clearly can't help myself, so I'm going to unwatch the PR.

Philippe23 commented:
@juippis:

- Sibling directory doesn't work.  See my comment above: https://github.com/gentoo/gentoo/pull/14055#issuecomment-582916443
- Thoughts on `/var/lib/spamd` vs `/etc/spamd`?  See mjo's last comment above regarding FHS:   https://github.com/gentoo/gentoo/pull/14055#issuecomment-582929503

`/var/lib/xxx` seems to be where almost all the other acct-user's are making homes.  Most other posts around the internet (ie: not Gentoo specific) that reference where they're putting spamd's home are also in `/var/lib/xxx` or `/home/xxx`.  The folder contains some configuration files and some book-keeping data (bayes toks + seen, and tx-reputation BerkDBs).  The configuration files are the root of mjo's FHS complaints.

Figured I'd get your input, since you're likely to be the one accepting it.

Frankly, even though it's a violation of the FHS, I'm leaning toward `/var/lib/spamd`, since `/etc/spamd` just seems wrong for the BerkDBs.  But I'm open to other alternatives.

juippis commented on acct-group/spamd/metadata.xml:
```diff
@@ -0,0 +1,18 @@
+  <maintainer type="person">
+    <email>bug@mejor.pl</email>
+    <name>Marcin Mirosław</name>
+  </maintainer>
```
Who is this? Do you have their consent on being added as a maintainer?

Philippe23 replied:
Marcin is co-proxy maintainer of spamassassin:
 https://bugs.gentoo.org/583908

But I have not gotten his explicit consent for the user/group packages.

juippis commented:
If the `/var/lib` works, I don't see any problem with it.

Philippe23 commented:
`/var/lib/spamd` functions - I've been using it since the last force push.

Let me know your 2¢ on the contents of that new text block at the bottom.  Especially with regards to verbage/clairity and that conversation on the dev mailing list about "user eclass should never change existing accounts" since as written those instructions lean heavily on the fact that the current behavior does.

juippis commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+	use postgres && dodoc sql/*_pg.sql
+	use mysql && dodoc sql/*_mysql.sql
+
+	dodoc NOTICE TRADEMARK CREDITS UPGRADE USAGE sql/README.bayes \
+		sql/README.awl procmailrc.example sample-nonspam.txt \
+		sample-spam.txt spamd/PROTOCOL spamd/README.vpopmail \
+		spamd-apache2/README.apache
+
+	# Rename some files so that they don't clash with others.
+	newdoc spamd/README README.spamd
+	newdoc sql/README README.sql
+	newdoc ldap/README README.ldap
```
You may want to use `DOCS` array so these get installed with `einstalldocs` like they're supposed to.

Philippe23 replied:
I did this for the `dodoc` block, but I left the renames as I'm not aware that `DOCS` has support for renaming.  If I'm wrong, please point me to some documentation.

juippis commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+	fi
+
+	# Remove perllocal.pod to avoid file collisions (bug #603338).
+	perl_delete_localpod || die "failed to remove perllocal.pod"
```
Eclass helper function, no `|| die` needed.

Philippe23 replied:
Removed.

juippis commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+	# perl_delete_emptybsdir and perl_remove_temppath, don't seem
+	# to be needed: there are no empty directories, *.bs files, or
+	# ${D} paths remaining in our installed image.
+	perl_fix_packlist || die "failed to fix paths in packlist"
```
Here too.

Philippe23 replied:
Removed.

Philippe23 commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+#
+# We still need the old Digest-SHA1 because razor2 has not been ported
+# to Digest-SHA.
+OPTDEPEND="app-crypt/gnupg
```
I think this is more about cross-compilation than it is about optional vs required.  I believe the name is just legacy from how they originally discovered what was needed on the building machine vs the target machine.

Philippe23 commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
```
I'd like to keep EAPI=7 in a separate revision so any bugs that arise can be traced to which change caused them.

Philippe23 commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+		dev-perl/DBI
+		dev-perl/DBD-Pg
+	)
+	sqlite? (
```
I'm not sure I follow.  Enabling the `sqlite` use flag will install the dependencies required by that database so it will function with that backend.

juippis replied:
Because `+sqlite` or `-sqlite` won't make any difference of the compiled product, it is useless to compile this program again whether that flag changes. It's banned by QA, see 
https://projects.gentoo.org/qa/policy-guide/dependencies.html#pg0001
(unless the build system has some automatic detection whether it's available or not - then you need to patch it to make it conditional)

kentfredric replied:
Well, being perl, it may have some sort of runtime detection, or alternatively, a path users can sometimes use on the installed code, but without the dependency, will have a fatal error.

Thus, depending on how often its needed, you'll want it in RDEPEND for some userbase.

But some users will not want the added weight and complexity the dependency produces, which in the case of DBD-SQLite, involves compiling a copy of SQLite ( because it can't use the system sqlite reliably )

This is really a place where there are no good options currently.

Based on the user needing to install the right SA to work, it seems like this situation in perl could fall under "such as a package being nonfunctional unless at least one of a set of optional runtime dependencies is installed."

Philippe23 replied:
This setup has existed since at least 2006.  (I see it in the 3.1.1 ebuild, it probably pre-dates that even.)  Changing it may break a lot of people's setups.  I think in this case it makes more sense to leave it as it is rather than expect users to have to manually manage a ton of dependencies just because Perl happens to be an interpreted language rather than a compiled one.

juippis replied:
Okay, thanks for explaining.

Philippe23 commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+
+	dosym mail/spamassassin /etc/spamassassin
+
+	# Disable plugin by default
+	sed -i -e 's/^loadplugin/\#loadplugin/g' \
+		"${ED}/etc/mail/spamassassin/init.pre" \
+		|| die "failed to disable plugins by default"
```
Moved.

Philippe23 commented on mail-filter/spamassassin/spamassassin-3.4.4-r1.ebuild:
```diff
@@ -0,0 +1,321 @@
+
+	systemd_newunit "${FILESDIR}/${PN}.service-r4" "${PN}.service"
+	systemd_install_serviced "${FILESDIR}/${PN}.service.conf-r2" \
+							 "${PN}.service"
```
Fixed.

Philippe23 commented on acct-user/spamd/metadata.xml:
```diff
@@ -0,0 +1,18 @@
+  <maintainer type="person">
+    <email>bug@mejor.pl</email>
+    <name>Marcin Mirosław</name>
+  </maintainer>
```
Removed.  I'll ping him and let him know he should make a PR to add himself if he wishes.

Philippe23 commented:
> 
> 
> So sorry for forgetting this, it's been an open tab with unfinished review in my browser for months. I hope you're now willing to push it through.

@juippis 

Most changes made, the others I replied with why I didn't (or at least why I feel they should be a separate revision).

Oh, and I wanted to make sure you saw my comment from last time:

> Let me know your 2¢ on the contents of that new text block at the bottom. Especially with regards to verbage/clairity and that conversation on the dev mailing list about "user eclass should never change existing accounts" since, as written those instructions lean heavily on the fact that the current behavior does.

I'm always paranoid that I write confusing docs.

juippis commented:
I think it's better to play it safe and be verbose for now. Can be updated later on.

Philippe23 commented:
> Sorry it took this long.

@juippis, no worries.  It's been less than straight-forward one.  Thanks again for all your work on this.  There'll be more PR's on their way soon (starting with EAPI=7).

Now that this change is in, the UID/GID should get marked in the API repo: https://github.com/gentoo/api-gentoo-org/pull/245
